sound/hda/intel-dsp-config.c

```diff
@@ -368,6 +417,12 @@ int snd_intel_dsp_driver_probe(struct pci_dev *pci)
 		return SND_INTEL_DSP_DRIVER_ANY;
 
 	if (cfg->flags & FLAG_SOF) {
+		if (cfg->flags & FLAG_SOF_ONLY_IF_SOUNDWIRE) {
+			if (snd_intel_dsp_check_soundwire(pci) > 0) {
+				dev_info(&pci->dev, "SoundWire enabled on CannonLake+ platform, using SOF driver\n");
+				return SND_INTEL_DSP_DRIVER_SOF;
+			}
+		}
```

Comment:
The logic here seems wrong. If we have `cfg->flags & FLAG_SOF_ONLY_IF_SOUNDWIRE == FLAG_SOF_ONLY_IF_SOUNDWIRE` but then `snd_intel_dsp_check_soundwire()` returns `<=0` that should mean, that we should fall back to SST or LEGACY. Instead we'll test for `FLAG_SOF_ONLY_IF_DMIC` below and if that isn't set, we'll return SOF, which is wrong.
What should the logic of these flags be? Can there be a situation where we want either `SOUNDWIRE` or `DMIC` or only both? Depending on the answer to this question the correct logic here would be different.

Reply:
@plbossart Ack, this should have a similar else branch as the ONLY_IF_DMIC check.

Hmm, this will get more complicated if we add the ONLY_IF_NON_HDA_CODECS flag as well.

Reply:
yes, this is wrong, but it can be fixed with just an else added to line 426

Reply:
@lyakh we want to select SOF wihen one or more of SoundWire/DMIC is set.

Reply:
> @lyakh we want to select SOF wihen one or more of SoundWire/DMIC is set.

@plbossart then I think this is what would work here:
```
	if (cfg->flags & FLAG_SOF_ONLY_IF_SOUNDWIRE && snd_intel_dsp_check_soundwire(pci) > 0) {
		dev_info(&pci->dev, "SoundWire enabled on CannonLake+ platform, using SOF driver\n");
		return SND_INTEL_DSP_DRIVER_SOF;
	}
	if (cfg->flags & FLAG_SOF_ONLY_IF_DMIC && snd_intel_dsp_check_dmic(pci)) {
		dev_info(&pci->dev, "Digital mics found on Skylake+ platform, using SOF driver\n");
		return SND_INTEL_DSP_DRIVER_SOF;
	}
	if (!(cfg->flags & (FLAG_SOF_ONLY_IF_DMIC | FLAG_SOF_ONLY_IF_SOUNDWIRE)))
		return SND_INTEL_DSP_DRIVER_SOF;
```

Reply:
> @lyakh we want to select SOF wihen one or more of SoundWire/DMIC is set.

@plbossart then I think this is what would work here:
```
	if (cfg->flags & FLAG_SOF_ONLY_IF_SOUNDWIRE && snd_intel_dsp_check_soundwire(pci) > 0) {
		dev_info(&pci->dev, "SoundWire enabled on CannonLake+ platform, using SOF driver\n");
		return SND_INTEL_DSP_DRIVER_SOF;
	}
	if (cfg->flags & FLAG_SOF_ONLY_IF_DMIC && snd_intel_dsp_check_dmic(pci)) {
		dev_info(&pci->dev, "Digital mics found on Skylake+ platform, using SOF driver\n");
		return SND_INTEL_DSP_DRIVER_SOF;
	}
	if (!(cfg->flags & (FLAG_SOF_ONLY_IF_DMIC | FLAG_SOF_ONLY_IF_SOUNDWIRE)))
		return SND_INTEL_DSP_DRIVER_SOF;
```